lib/backends/webgl/ops/upsample.ts

```diff
@@ -0,0 +1,193 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+import {Upsample} from '../../../ops/upsample';
+import {Tensor} from '../../../tensor';
+import {getGlsl} from '../glsl-source';
+import {WebGLInferenceHandler} from '../inference-handler';
+import {ProgramInfo, RunData, WebGLOperator} from '../types';
+
+export class WebGLUpsample extends Upsample implements WebGLOperator {
+  run(inferenceHandler: WebGLInferenceHandler, inputs: Tensor[]): Tensor[] {
+    return inferenceHandler.run(this, inputs);
+  }
+  createProgramInfo(handler: WebGLInferenceHandler, inputs: Tensor[]): ProgramInfo {
+    const inputLayout = handler.getOrCreateTextureLayout(inputs[0]);
+    const outputShape = inputs[0].dims.map((dim, i) => Math.floor(dim * this.scales[i]));
+    const outputLayout = handler.createTextureLayoutFromShape(outputShape);
+    const dim = outputShape.length;
+
+    const glsl = getGlsl(handler.session.backend.glContext.version);
+
+    const outputPitches = new Array<number>(dim);
+    const inputPitches = new Array<number>(dim);
+    let precalculatedPitches = `
+      int output_pitches[${dim}];
+      int input_pitches[${dim}];
+      `;
+    for (let d = dim - 1; d >= 0; d--) {
+      outputPitches[d] = (d === dim - 1) ? 1 : outputPitches[d + 1] * outputShape[d + 1];
```

Comment:
Set outputPitches[dim - 1] and inputPitches[dim - 1] before this loop to avoid unnecessary comparison.

Reply:
I did this change locally but then I realize that this requires to duplicate the 4 lines of code in the for body, which makes the code hard to read. considering this condition expression does very little impact to performance, and this code only happens in initialization, it does not have enough benefit to balance the lost of readability; so I will not change this place.